<a href="https://colab.research.google.com/github/harshamule/CTR-Prediction-ML/blob/main/Diabetes_Prediction_ML.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [3]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.linear_model import LogisticRegression

In [4]:
df=pd.read_csv("/content/diabetes datset.zip")
df

,Pregnancies,Glucose,BloodPressure,SkinThickness,Insulin,BMI,DiabetesPedigreeFunction,Age,Outcome
0,6,148,72,35,0,33.6,0.627,50,1
1,1,85,66,29,0,26.6,0.351,31,0
2,8,183,64,0,0,23.3,0.672,32,1
3,1,89,66,23,94,28.1,0.167,21,0
4,0,137,40,35,168,43.1,2.288,33,1
...,...,...,...,...,...,...,...,...,...
763,10,101,76,48,180,32.9,0.171,63,0
764,2,122,70,27,0,36.8,0.340,27,0
765,5,121,72,23,112,26.2,0.245,30,0
766,1,126,60,0,0,30.1,0.349,47,1


In [5]:
df.shape

(768, 9)

In [6]:
df.isnull().sum()

,0
Pregnancies,0
Glucose,0
BloodPressure,0
SkinThickness,0
Insulin,0
BMI,0
DiabetesPedigreeFunction,0
Age,0
Outcome,0


In [7]:
df_dup=df.duplicated()
df_dup

,0
0,False
1,False
2,False
3,False
4,False
...,...
763,False
764,False
765,False
766,False


In [8]:
df.drop_duplicates(inplace=True)
df.duplicated().sum()



np.int64(0)

In [9]:
df.shape

(768, 9)

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Pregnancies               768 non-null    int64  
 1   Glucose                   768 non-null    int64  
 2   BloodPressure             768 non-null    int64  
 3   SkinThickness             768 non-null    int64  
 4   Insulin                   768 non-null    int64  
 5   BMI                       768 non-null    float64
 6   DiabetesPedigreeFunction  768 non-null    float64
 7   Age                       768 non-null    int64  
 8   Outcome                   768 non-null    int64  
dtypes: float64(2), int64(7)
memory usage: 54.1 KB


In [11]:
import plotly.express as px


In [12]:
col_list=df.columns

In [13]:
for i in col_list:
  if df[i].dtype!='object':
    fig=px.box(df[i],title=f"box plot of{i}column")
    fig.show()


In [14]:
df.columns

Index(['Pregnancies', 'Glucose', 'BloodPressure', 'SkinThickness', 'Insulin',
       'BMI', 'DiabetesPedigreeFunction', 'Age', 'Outcome'],
      dtype='object')

In [15]:
out_liers=['Pregnancies','BloodPressure','Insulin','BMI', 'DiabetesPedigreeFunction', 'Age']
for col in out_liers:
  q1=df[col].quantile(0.25)
  q3=df[col].quantile(0.75)
  iqr=q3-q1
  ul=q3+1.5*iqr
  ll=q1-1.5*iqr
  df=df[(df[col]>=ll)&(df[col]<=ul)]

In [16]:
for i in col_list:
  if df[i].dtype!='object':
    fig=px.box(df[i],title=f"box plot of{i}column")
    fig.show()


In [17]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 640 entries, 0 to 767
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Pregnancies               640 non-null    int64  
 1   Glucose                   640 non-null    int64  
 2   BloodPressure             640 non-null    int64  
 3   SkinThickness             640 non-null    int64  
 4   Insulin                   640 non-null    int64  
 5   BMI                       640 non-null    float64
 6   DiabetesPedigreeFunction  640 non-null    float64
 7   Age                       640 non-null    int64  
 8   Outcome                   640 non-null    int64  
dtypes: float64(2), int64(7)
memory usage: 50.0 KB


In [18]:
df.shape

(640, 9)

feature engineering


In [19]:
from sklearn.model_selection import train_test_split

x = df.drop('Outcome',axis=1)  # axis=1 tells pandas to drop a column entire, not an index
y = df["Outcome"]
x

,Pregnancies,Glucose,BloodPressure,SkinThickness,Insulin,BMI,DiabetesPedigreeFunction,Age
0,6,148,72,35,0,33.6,0.627,50
1,1,85,66,29,0,26.6,0.351,31
2,8,183,64,0,0,23.3,0.672,32
3,1,89,66,23,94,28.1,0.167,21
5,5,116,74,0,0,25.6,0.201,30
...,...,...,...,...,...,...,...,...
763,10,101,76,48,180,32.9,0.171,63
764,2,122,70,27,0,36.8,0.340,27
765,5,121,72,23,112,26.2,0.245,30
766,1,126,60,0,0,30.1,0.349,47


In [20]:
y

,Outcome
0,1
1,0
2,1
3,0
5,0
...,...
763,0
764,0
765,0
766,1


In [21]:
x_train,x_test,y_train,y_test=train_test_split(x,y,train_size=0.8,random_state=0)


In [22]:
model=LogisticRegression(max_iter=1000)
model.fit(x_train,y_train)

LogisticRegression(max_iter=1000)

In [23]:
y_pred=model.predict(x_test)
y_pred

array([0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1, 0, 1, 0,
       0, 0, 1, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0,
       1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 1, 1, 0, 1,
       1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 1,
       1, 1, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0])

In [24]:
from sklearn.metrics import accuracy_score,confusion_matrix,classification_report,r2_score
accuracy_score(y_test,y_pred)


0.8125

In [25]:
confusion_matrix(y_test,y_pred)


array([[83,  8],
       [16, 21]])

In [26]:
print(classification_report(y_test,y_pred))


              precision    recall  f1-score   support

           0       0.84      0.91      0.87        91
           1       0.72      0.57      0.64        37

    accuracy                           0.81       128
   macro avg       0.78      0.74      0.76       128
weighted avg       0.81      0.81      0.81       128



In [27]:
r2_score(y_test,y_pred)

0.08761508761508763

In [28]:
from sklearn.model_selection import cross_val_score
scores = cross_val_score(model, x, y, cv=5, scoring='accuracy')
print("Mean Accuracy:", scores.mean())

Mean Accuracy: 0.7796875


decison tree


In [29]:
from sklearn.tree import DecisionTreeClassifier
model_dt=DecisionTreeClassifier()

In [30]:
model_dt.fit(x_train,y_train)

DecisionTreeClassifier()

In [31]:
y_pred_dt=model_dt.predict(x_test)

In [32]:
y_pred_dt

array([0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0, 1, 0, 0, 0, 1, 0, 0, 1, 0, 1, 1,
       0, 0, 0, 1, 0, 1, 1, 1, 0, 1, 0, 0, 0, 0, 1, 0, 1, 1, 0, 0, 1, 0,
       1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 1,
       1, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0,
       0, 1, 0, 1, 0, 0, 0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0])

In [33]:
accuracy_score(y_test,y_pred_dt)

0.6875

In [34]:
from sklearn.ensemble import RandomForestClassifier
model_rf=RandomForestClassifier()

In [35]:
model_rf.fit(x_train,y_train)

RandomForestClassifier()

In [36]:
y_pred_rf=model_rf.predict(x_test)
y_pred_rf

array([0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 1, 0, 1, 1,
       0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1,
       1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 1, 1, 0, 1,
       1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 1, 0, 0, 0, 0, 0, 1,
       1, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0])

In [37]:
accuracy_score(y_test,y_pred_rf)

0.8046875

hyperameter tuning


In [38]:
from sklearn.model_selection import GridSearchCV

🎯 Use of GridSearchCV

👉 The main use of GridSearchCV is to automatically find the best parameters (hyperparameters) for your machine learning model — in your case, the Random Forest Classifier.

🧠 Think of it like this:

When you build a model (like RandomForest), you have many settings you can change:

How many trees to grow (n_estimators)

How deep the trees can go (max_depth)

How many samples to split (min_samples_split)

etc.

These are called hyperparameters — and they greatly affect your model’s accuracy.
But you don’t know which combination works best.

So instead of manually guessing 👇

rf = RandomForestClassifier(n_estimators=100, max_depth=10)


you let GridSearchCV try many combinations automatically and pick the best one.

⚙️ What GridSearchCV does:

Takes all parameter combinations from the grid (you defined in param_grid).

For each combination:

It trains your model multiple times (cross-validation).

It tests how well the model performs.

Finds the combination that gives the highest accuracy (or any metric you choose).

📊 Example

Let’s say you test:

n_estimators	max_depth	Accuracy
100	10	0.78
200	20	0.82
500	30	0.80

GridSearchCV will automatically select:

Best Parameters: {'n_estimators': 200, 'max_depth': 20}
Best Accuracy: 0.82


Now you can build your final model with these best settings to get maximum accuracy and stability.

💡 So in simple words:
Without GridSearchCV	With GridSearchCV
You guess the parameters manually	It tests all combinations for you
Risk of underfitting or overfitting	Finds best balance automatically
One model trained	Many models tested
Lower accuracy	Optimized accuracy

✅ Final Use:

To automatically tune and select the best hyperparameters that make your model perform better and more reliable on unseen data.

In [40]:
"""paramgrid={
    'n_estimators':[100,200,300,400,500],
    'max_depth':[None,10,20,30],
    'min_samples_split':[2,5,10],
    'min_samples_leaf':[1,2,4],
    'bootstrap':[True,False]
}

rf=RandomForestClassifier()
grid_search=GridSearchCV(estimator=rf,param_grid=paramgrid,cv=5,n_jobs=-1,verbose=2,scoring='accuracy')
grid_search.fit(x_train,y_train)"""

"paramgrid={\n    'n_estimators':[100,200,300,400,500],\n    'max_depth':[None,10,20,30],\n    'min_samples_split':[2,5,10],\n    'min_samples_leaf':[1,2,4],\n    'bootstrap':[True,False]\n}\n\nrf=RandomForestClassifier()\ngrid_search=GridSearchCV(estimator=rf,param_grid=paramgrid,cv=5,n_jobs=-1,verbose=2,scoring='accuracy')\ngrid_search.fit(x_train,y_train)"

In [41]:
#y_pred_grid=grid_search.predict(x_test)

In [42]:
#accuracy_score(y_test,y_pred_grid)

gradient boosting


In [43]:
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

In [44]:
gb = GradientBoostingClassifier(
    n_estimators=300,       # number of boosting stages
    learning_rate=0.1,      # step size shrinkage
    max_depth=2,            # depth of individual estimators
    random_state=42
)

# Train model
gb.fit(x_train, y_train)

# Predict on test data
y_pred = gb.predict(x_test)

# Evaluate performance
acc = accuracy_score(y_test, y_pred)
print("✅ Accuracy:", acc)
print("\nConfusion Matrix:\n", confusion_matrix(y_test, y_pred))
print("\nClassification Report:\n", classification_report(y_test, y_pred))

✅ Accuracy: 0.7890625

Confusion Matrix:
 [[80 11]
 [16 21]]

Classification Report:
               precision    recall  f1-score   support

           0       0.83      0.88      0.86        91
           1       0.66      0.57      0.61        37

    accuracy                           0.79       128
   macro avg       0.74      0.72      0.73       128
weighted avg       0.78      0.79      0.78       128



testing it on new data

enter the patient values dyanamic values

In [48]:
 """import pandas as pd

def predict_new_patient(model):
    while True:
        print("\nEnter new patient data:")

        # Take dynamic input for each feature
        try:
            Pregnancies = int(input("Pregnancies: "))
            Glucose = float(input("Glucose: "))
            BloodPressure = float(input("BloodPressure: "))
            SkinThickness = float(input("SkinThickness: "))
            Insulin = float(input("Insulin: "))
            BMI = float(input("BMI: "))
            DiabetesPedigreeFunction = float(input("DiabetesPedigreeFunction: "))
            Age = int(input("Age: "))
        except ValueError:
            print("⚠️ Invalid input! Please enter numeric values only.")
            continue

        # Create DataFrame for the new patient
        new_data = pd.DataFrame([{
            'Pregnancies': Pregnancies,
            'Glucose': Glucose,
            'BloodPressure': BloodPressure,
            'SkinThickness': SkinThickness,
            'Insulin': Insulin,
            'BMI': BMI,
            'DiabetesPedigreeFunction': DiabetesPedigreeFunction,
            'Age': Age
        }])

        # Predict outcome
        prediction = gb.predict(new_data)
        print(f"🧾 Predicted Outcome: {prediction[0]} (0 = No Diabetes, 1 = Diabetes)")

        # Ask if user wants to test another patient
        again = input("\nDo you want to enter another patient? (yes/no): ").strip().lower()
        if again != 'yes':
            print("✅ Exiting dynamic prediction.")
            break

# Usage
# Replace best_gb with your trained Gradient Boosting model
predict_new_patient(gb)"""


'import pandas as pd\n\ndef predict_new_patient(model):\n   while True:\n       print("\nEnter new patient data:")\n\n       # Take dynamic input for each feature\n       try:\n           Pregnancies = int(input("Pregnancies: "))\n           Glucose = float(input("Glucose: "))\n           BloodPressure = float(input("BloodPressure: "))\n           SkinThickness = float(input("SkinThickness: "))\n           Insulin = float(input("Insulin: "))\n           BMI = float(input("BMI: "))\n           DiabetesPedigreeFunction = float(input("DiabetesPedigreeFunction: "))\n           Age = int(input("Age: "))\n       except ValueError:\n           print("⚠️ Invalid input! Please enter numeric values only.")\n           continue\n\n       # Create DataFrame for the new patient\n       new_data = pd.DataFrame([{\n           \'Pregnancies\': Pregnancies,\n           \'Glucose\': Glucose,\n           \'BloodPressure\': BloodPressure,\n           \'SkinThickness\': SkinThickness,\n           \'Insulin

In [50]:
print("Logistic Regression:", accuracy_score(y_test,y_pred))
print("Decision Tree:", accuracy_score(y_test,y_pred_dt))
print("Gradient Boosting:", accuracy_score(y_test,y_pred))
print("Random Forest:", accuracy_score(y_test,y_pred_rf))



Logistic Regression: 0.7890625
Decision Tree: 0.6875
Gradient Boosting: 0.7890625
Random Forest: 0.8046875


1. Pregnancies

How many times a woman has been pregnant.

More pregnancies can slightly increase diabetes risk.

2. Glucose

Sugar level in the blood.

Higher sugar → higher chance of diabetes. This is the most important factor.

3. BloodPressure

The force of blood in the arteries.

Higher blood pressure can be linked to diabetes risk.

4. SkinThickness

How thick the fat under the skin is.

Higher fat → higher risk.

5. Insulin

Amount of insulin in the blood.

Very high or low → indicates problems controlling sugar.

6. BMI

Body Mass Index — how heavy someone is for their height.

Higher BMI → higher diabetes risk.

7. DiabetesPedigreeFunction

Shows if diabetes runs in the family.

Higher value → higher genetic risk.

8. Age

Older people are more likely to get diabetes.

9. Outcome

0 = No Diabetes, 1 = Diabetes (this is what we’re predicting).
| Feature                      | Typical Range    | Notes / Inference                                                                      |
| ---------------------------- | ---------------- | -------------------------------------------------------------------------------------- |
| **Pregnancies**              | 0 – 17           | Usually 0–5 is common. More pregnancies slightly increase risk.                        |
| **Glucose**                  | 70 – 200         | 70–99 = normal, 100–125 = prediabetes, >125 = diabetes risk. Most important predictor. |
| **BloodPressure**            | 40 – 122 mmHg    | Normal ~60–80. High blood pressure increases risk.                                     |
| **SkinThickness**            | 7 – 99 mm        | Higher values indicate more body fat → higher risk.                                    |
| **Insulin**                  | 15 – 276 mu U/ml | Very low or very high insulin can indicate diabetes.                                   |
| **BMI**                      | 18 – 67          | Normal 18–24, overweight 25–29, obese 30+ → higher risk.                               |
| **DiabetesPedigreeFunction** | 0.08 – 2.42      | Higher values → stronger family history → higher risk.                                 |
| **Age**                      | 21 – 81          | Risk increases with age, especially 35+.                                               |
| **Outcome**                  | 0 or 1           | 0 = No Diabetes, 1 = Diabetes (target column).                                         |



 | Feature                  | Healthy Patient | Risky Patient |
| ------------------------ | --------------- | ------------- |
| Pregnancies              | 1               | 5             |
| Glucose                  | 90              | 150           |
| BloodPressure            | 70              | 85            |
| SkinThickness            | 20              | 35            |
| Insulin                  | 80              | 200           |
| BMI                      | 23              | 32            |
| DiabetesPedigreeFunction | 0.3             | 1.2           |
| Age                      | 28              | 50            |
| Outcome                  | 0 (No Diabetes) | 1 (Diabetes)  |
General Inferences from the Dataset

Glucose is the strongest predictor — high glucose almost always leads to diabetes.

BMI, Age, Insulin, and DiabetesPedigreeFunction are secondary predictors — they refine the risk assessment.

Pregnancies and BloodPressure have smaller but still notable effects.

Features interact: For example, a patient with moderate BMI but high glucose and strong family history may still be predicted as diabetic.

Imbalanced classes: Usually, fewer patients have diabetes than don’t — this affects precision/recall.
High BMI + high glucose → strong risk → model predicts 1.
High-risk patients: Glucose > 125, BMI > 30, Age > 40, strong family history → Outcome = 1

Low-risk patients: Glucose < 100, BMI < 28, younger age, low family history → Outcome = 0